src/cpp/src/visual_language/input_embedder.cpp

```diff
@@ -461,73 +461,107 @@ class InputsEmbedderLLaVA : public InputsEmbedder::IInputsEmbedder {
 
     virtual ov::Tensor get_inputs_embeds(const std::string& prompt, const std::vector<ov::Tensor>& images) override {
         std::string image_token = m_vlm_config.im_start;
-        std::string formatted_prompt = images.empty() ? prompt : image_token + "\n" + prompt;
-
-        // std::string chat_template_fallback = m_templated_chat_history + " USER: " + formatted_prompt + " ASSISTANT: ";
-        // chat_template_fallback = chat_template_fallback.erase(0, chat_template_fallback.find_first_not_of(' '));
-
         // Adapted from llava-1.5-7b-hf chat_template.json
         std::string chat_template_fallback = "{% for message in messages %}{% if message['role'] == 'user' %}{{ 'USER: ' + message['content'] + ' ' }}{% else %}{{ 'ASSISTANT: ' + message['content'] + ' ' }}{% endif %}{% endfor %}{% if add_generation_prompt %}{{ 'ASSISTANT:' }}{% endif %}";
+        
+        std::string formatted_prompt;
+        std::vector<ov::Tensor> image_embeds;
+        image_embeds.reserve(images.size());
+
+        for (const auto& image : images) {
```

Comment:
Could you factor this loop into a function to avoid duplication? And here  you just build `formatted_prompt`.

Reply:
For llava models this loop not only builds `formatted_prompt`, but also populates `image_embeds` after encoding images.
For llava-next models the loop is similar, but it additionally processes encoded image (e.g. packing image features, adding image newline tensor) before adding to `image_embeds` vector.

The loop can be moved to separate function (e.g. `encode_images()`) returning encoded images vector, but this will require one more looping over encoded images vector to get image embeds (this logic is different for llava and llava-next use cases) and populating `formatted_prompt`, e.g.
```
std::vector<ov::Tensor> image_embeds;
...
std::vector<EncodedImage> encoded_images = encode_images(images);
for (const auto& encoded_image : encoded_images ) {
    // getting image embeds
    ...
    image_embeds.push_back(std::move(single_image_embeds));
    formatted_prompt += image_token + "\n";
}
```